packages/sync/src/conflicts/conflictLink.ts

```diff
@@ -1,15 +1,152 @@
+ */
+export class LocalConflictError extends Error {
+  constructor(private base: any, private variables: any) {
+    super();
+  }
+}
+
+/**
+ * Represents conflict information that was returned from server
+ */
+export interface ConflictInfo {
+  serverState: ConflictResolutionData
+  clientState: ConflictResolutionData
+  // Expected return type of the mutation
+  returnType: string
+}
+
+/**
+ * Configuration for conflict resolution
+ */
+export interface ConflictConfig {
+  /**
+   * Interface that defines how object state is progressed
+   * This interface needs to match state provider supplied on server.
+   */
+  conflictProvider: ObjectState;
+
+  /**
+   * Interface that can be implemented to receive information about the data conflict
+   *
+   * @deprecated see OfflineClient.registerOfflineEventListener
+   */
+  conflictListener?: ConflictListener;
+
+  /**
+   * The conflict resolution strategy your client should use. By default it takes client version.
+   */
+  conflictStrategy?: ConflictResolutionStrategies;
+
+  /**
+   * Base state provider gives ability to save base state
+   */
+  baseState: BaseStateProvider
+}
+
+/**
+ * Conflict handling link implementation that provides ability to determine 
+ */
+export class ConflictLink extends ApolloLink {
+  private link: ApolloLink;
+  private stater: ObjectState;
+  private strategy: ConflictResolutionStrategies | undefined;
+  private listener: ConflictListener | undefined;
+
+  constructor(private config: ConflictConfig, private cache: InMemoryCache) {
+    super();
+    this.link = onError((errorResponse) => {
+      this.conflictErrorHandler(errorResponse);
+    });
+    this.stater = this.config.conflictProvider;
+    this.strategy = this.config.conflictStrategy;
+    this.listener = this.config.conflictListener;
+  }
+
+  public request(
+    operation: Operation,
+    forward: NextLink,
+  ): Observable<FetchResult> | null {
+    if (!isMutation(operation)) {
+      // 🙈 Nothing to do here 
+      return forward(operation);
+    }
+    const currentState = this.stater.currentState(operation.variables);
+    if (!currentState) {
+      // 🙈 Missing state information in request.
+      // No action here
+      return forward(operation);
+    }
+    this.processBaseState(operation, forward);
+    // FIXME removing base state
+    return this.link.request(operation, forward);
+  }
+
+  private processBaseState(operation: Operation, forward: NextLink) {
+    const context = operation.getContext();
+    const base = getObjectFromCache(this.cache, operation.variables.id, context.returnType);
+    if (base && Object.keys(base).length != 0) {
+      const currentChange = this.stater.currentState(base);
+      const userState = this.stater.currentState(operation.variables);
+      if (currentChange !== userState) {
+         // 🙊 Input data is conflicted with the latest server projection
+        throw new LocalConflictError(base, operation.variables);
```

Comment:
This is actually not going to work since we should be returning observable with error. 
THrowing error here breaks js stack